### DISCUSSION TITLE
Automatically refreshing credentials with STS `AssumeRole` for API Gateway request signing

### DISCUSSION THREAD

**Discussion by avandesa-fw** [awslabs/aws-sdk-rust#1177] [Q&A] [answered]
Hello! We have a somewhat odd use case and I'm looking into ways to leverage the SDK to make our lives easier.

Currently, we're calling [`AssumeRole`](https://docs.rs/aws-sdk-sts/latest/aws_sdk_sts/struct.Client.html#method.assume_role) to get temporary credentials, then using those to manually sign HTTP requests with [`aws_sigv4`](https://docs.rs/aws-sigv4/latest/aws_sigv4/) to authenticate against an AWS API Gateway. This works great so far! However, we need some way to refresh those credentials since they only last a couple hours.

Our tentative plan is to keep the credentials behind a lock of some sort, then re-fetch them when requests start failing with `403 Forbidden` errors. However, this synchronization is going to be tough to get right. I was wondering if there's any existing tools in the SDK or smithy to make this process easier.

One possibility I see is giving an SDK client a [`ProvideCredentials` impl](https://docs.rs/aws-credential-types/1.2.0/aws_credential_types/provider/trait.ProvideCredentials.html) that internally calls `AssumeRole`, then getting the [`SharedHttpClient`](https://docs.rs/aws-sdk-sts/latest/aws_sdk_sts/config/struct.SharedHttpClient.html) and using that to perform requests, but it's not immediately clear to me if
* the client would only call `ProvideCredentials::provide_credentials` when the creds expire
* the `SharedHttpClient` or `SharedHttpConnector` would have access to those creds, or
* we'd be able to configure the other signing parameters, like the region and service.

Again, we can do the synchronization ourselves, but I want to make sure I'm not missing some other SDK feature that would make this easier. Thanks!

**Comment by ysaito1001** [comment, marked as answer]
Thanks for posting! Normally, you'd reach out something like [AssumeRoleProvider](https://docs.rs/aws-config/latest/aws_config/sts/struct.AssumeRoleProvider.html), create an instance of it, and pass it to the [credentials_provider](https://docs.rs/aws-config/latest/aws_config/struct.ConfigLoader.html#method.credentials_provider) method on `ConfigLoader`. Once you constructed a service client from `ConfigLoader`, then that `AssumeRoleProvider` would be backed by an identity cache (ex. [lazy identity cache](https://docs.rs/aws-config/latest/aws_config/identity/struct.LazyCacheBuilder.html)). This cache keeps track of credentials' expiry and it automatically refreshes credentials when they're about to expire.

Do you use `ConfigLoader` to create a service client? If not, how do you construct a service client currently?

**Reply by jmklix** [reply]
Marking above as the answer, please open a new discussion if you have any questions
